x-pack/plugins/lens/common/locator/locator.ts

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import rison from '@kbn/rison';
+import type { SerializableRecord } from '@kbn/utility-types';
+import type { GlobalQueryStateFromUrl } from '@kbn/data-plugin/public';
+import type { LocatorDefinition, LocatorPublic } from '@kbn/share-plugin/common';
+import type { Filter, Query } from '@kbn/es-query';
+import type { DataViewSpec, SavedQuery } from '@kbn/data-plugin/common';
+import { SavedObjectReference } from '@kbn/core-saved-objects-common';
+import type { DateRange } from '../types';
+
+export const LENS_APP_LOCATOR = 'LENS_APP_LOCATOR';
+export const LENS_SHARE_STATE_ACTION = 'LENS_SHARE_STATE_ACTION';
+
+interface LensShareableState {
+  /**
+   * Optionally apply filters.
+   */
+  filters?: Filter[];
+
+  /**
+   * Optionally set a query.
+   */
+  query?: Query;
+
+  /**
+   * Optionally set the date range in the date picker.
+   */
+  resolvedDateRange?: DateRange & SerializableRecord;
+
+  /**
+   * Optionally set the id of the used saved query
+   */
+  savedQuery?: SavedQuery & SerializableRecord;
+
+  /**
+   * Set the visualization configuration
+   */
+  visualization: { activeId: string | null; state: unknown } & SerializableRecord;
+
+  /**
+   * Set the active datasource used
+   */
+  activeDatasourceId?: string;
+
+  /**
+   * Set the datasources configurations
+   */
+  datasourceStates: Record<string, unknown> & SerializableRecord;
+
+  /**
+   * Background search session id
+   */
+  searchSessionId?: string;
+
+  /**
+   * Set the references used in the Lens state
+   */
+  references: Array<SavedObjectReference & SerializableRecord>;
+
+  /**
+   * Pass adHoc dataViews specs used in the Lens state
+   */
+  dataViewSpecs?: DataViewSpec[];
+}
+
+export interface LensAppLocatorParams extends SerializableRecord {
+  /**
+   * Optionally set saved object ID.
+   */
+  savedObjectId?: string;
+
+  /**
+   * Background search session id
+   */
+  searchSessionId?: string;
+
+  /**
+   * Optionally apply filters.
+   */
+  filters?: Filter[];
+
+  /**
+   * Optionally set a query.
+   */
+  query?: Query;
+
+  /**
+   * Optionally set the date range in the date picker.
+   */
+  resolvedDateRange?: DateRange & SerializableRecord;
+
+  /**
+   * Optionally set the id of the used saved query
+   */
+  savedQuery?: SavedQuery & SerializableRecord;
+
+  /**
+   * In case of no savedObjectId passed, the properties above have to be passed
+   */
```

Comment:
I think you could enforce this by way of a union type

```ts
export interface BaseLensAppLocatorParams extends SerializableRecord {
  /**
   * Set the active datasource used
   */
  activeDatasourceId?: string | null;

  /**
   * Set the visualization configuration
   */
  visualization?: { activeId: string | null; state: unknown } & SerializableRecord;

  /**
   * Set the datasources configurations
   */
  datasourceStates?: Record<string, { state: unknown }> & SerializableRecord;

  /**
   * Set the references used in the Lens state
   */
  references?: Array<SavedObjectReference & SerializableRecord>;

  /**
   * Pass adHoc dataViews specs used in the Lens state
   */
  dataViewSpecs?: DataViewSpec[];
}

type LensAppLocatorParams =
  | (BaseLensAppLocatorParams & {
      savedObjectId: string;
    })
  | {
      savedObjectId?: string;

      /**
       * Background search session id
       */
      searchSessionId: string;

      /**
       * Optionally apply filters.
       */
      filters: Filter[];

      /**
       * Optionally set a query.
       */
      query: Query;

      /**
       * Optionally set the date range in the date picker.
       */
      resolvedDateRange: DateRange & SerializableRecord;

      /**
       * Optionally set the id of the used saved query
       */
      savedQuery: SavedQuery & SerializableRecord;
    };
```

Reply:
Originally I thought about that, but one thing I do not like when consuming types is to not see the full type on hover or to go recursively to check the actual content of the type definition.